### ISSUE TITLE
may you add to  requirements.txt all needed packages

### ISSUE THREAD

**Issue by Sandy4321** [NirDiamant/RAG_Techniques#73]
may you add to [requirements.txt](https://github.com/NirDiamant/RAG_Techniques/blob/main/requirements.txt)  all needed packages 
for example 
for file 
https://github.com/NirDiamant/RAG_Techniques/blob/main/all_rag_techniques/graph_rag.ipynb
sklearn is not here
<img width="859" alt="image" src="https://github.com/user-attachments/assets/78d75117-835d-4cf4-a4c9-359e138a101e" />

 then which version for sklearn  to install ?

**Comment by Sandy4321** [comment]
my changes 
<img width="889" alt="image" src="https://github.com/user-attachments/assets/619fae13-86dc-4d18-b5a1-1ea03832b215" />

**Comment by Sandy4321** [comment]
#!pip install scikit-learn
#!pip install nltk
#!pip install spacy
#!pip install rank-bm25
#!pip install PyMuPDF
#!pip install deepeval

**Comment by Sandy4321** [comment]
it is what I get, pls check if it is correct result?
Processing query: what is the main cause of climate change?

Retrieving relevant documents...

Traversing the knowledge graph:

Step 1 - Node 2:
Content: driven by human activities, particularly the emission of greenhou se gases.  
Chapter 2: Causes of C...
Concepts: fossil fuels, industrial revolution, heating, Fossil Fuels, greenhouse effect, nitrous oxide, emission, CH4, fossil fuel consumption, climate change, human activities, atmosphere, coal, Coal, life on Earth, electricity, carbon dioxide, greenhou, methane, transportation, oil, warmer climate, natural gas, greenhouse gases, energy
--------------------------------------------------

Step 2 - Node 0 (neighbor of 2):
Content: Understanding Climate Change  
Chapter 1: Introduction to Climate Change  
Climate change refers to ...
Concepts: global climate, Earth's orbit, deforestation, wind patterns, precipitation, Understanding Climate Change  
, Holocene, historical context, climate change, human activities, solar energy, glacial advance, temperature, weather patterns, burning of fossil fuels, Holocene epoch, glacial retreat, ice age, human civilization, modern climate era
--------------------------------------------------

Step 3 - Node 1 (neighbor of 2):
Content: Most of these climate changes are attributed to very small variations in Earth's orbit that 
change ...
Concepts: Earth's orbit, The Intergovernmental Panel on Climate Change, future trends, past climate conditions, extreme weather events, industrial era, Holocene, climate change, human activities, solar energy, ocean sediments, global temperatures, Holocene epoch, ice core samples, greenhou, tree rings, historical record, Intergovernmental Panel on Climate Change, emission of greenhouse gases, sea levels
--------------------------------------------------

Step 4 - Node 3 (neighbor of 2):
Content: Coal is the most carbon -intensive fossil fuel, and its use for electricity generation is a major 
s...
Concepts: CO2 emissions, air quality, China, gasoline, carbon-intensive fossil fuel, economic implications, transportation fuels, the United 
States, global oil industry, extraction, climate change, electricity generation, geopolitical implications, greenhouse gas, bridge fuel, lower-carbon future, Coal, refining, mined, diesel, energy source, India, pollutants, distribution, natural gas, combustion
--------------------------------------------------

Step 5 - Node 4 (neighbor of 2):
Content: Natural gas is the least carbon -intensive fossil fuel and is often seen as a "bridge fuel" to a 
lo...
Concepts: greenhouse gas emissions, Innov, deforestation, tropical rainforests, greenhouse effect, health concerns, bridge fuel, Congo Basin, timber, global carbon cycles, lower-carbon future, Amazon, agriculture, Southeast Asia, carbon-intensive, carbon sinks, fossil fuel, CO2 absorption, methane, fracking technology, carbon storage, natural gas, environmental concerns
--------------------------------------------------

Step 6 - Node 5 (neighbor of 2):
Content: Tropical rainforests are particularly important for carbon storage. Deforestation in the 
Amazon, Co...
Concepts: deforestation, regulating climate, mining, indigenous communities, Livestock Emissions, livestock, wildlife, climate change, rice paddies, species extinction, heat-trapping capability, greenhouse gas, Congo Basin, Europe, habitat loss, global carbon cycles, Amazon, Tropical, Southeast Asia, agriculture, Asia, synthetic fertilizers, biodiversity, land-use changes, logging, Tropical rainforests, Boreal Forests, carbon storage, North America, sequestering carbon, methane emissions, CO2
--------------------------------------------------

Step 7 - Node 6 (neighbor of 2):
Content: Ruminant animals, such as cows and sheep, produce methane during digestion. Manure 
management pract...
Concepts: rice cultivation, methane production, Ruminant animals, manure management, water management, anaerobic conditions, climate change, flooded rice paddies, greenhouse gas, carbon footprint, agriculture, synthetic fertilizers, digestion, food security, farming techniques, eco-friendly fertilizers, waste management, livestock feeding, organic fertilizers, sustainable rice farming, climate goals, nitrous oxide emissions, rice varieties, precision farming
--------------------------------------------------

Step 8 - Node 7 (neighbor of 2):
Content: development of eco -friendly fertilizers and farming techniques is  essential for reducing the 
agri...
Concepts: ecosystems, urban heat island effect, effects of climate change, rising temperatures, carbon footprint, human health, changing seasons, global temperatures, heatwaves, farming techniques, eco-friendly fertilizers, heat-related illnesses, Rising Temperatures  
Global, agricultural sector, infrastructure
--------------------------------------------------

Step 9 - Node 8 (neighbor of 2):
Content: Changing Seasons  
Climate change is altering the timing and length of seasons, affecting ecosystems...
Concepts: ecosystems, agricultural practices, melting ice, Rising Sea Levels  
Warmer, spring, animal life cycles, polar ice caps, glaciers, coastal communities, climate change, human activities, plant life cycles, winters, seasons, rising sea levels
--------------------------------------------------

Step 10 - Node 9 (neighbor of 2):
Content: The Arctic is warming at more than twice the global average rate, leading to significant ice 
loss. ...
Concepts: Himalayas, Arctic warming, Coastal, glacial meltwater, storm surges, infrastructure, ice loss, ecosystems, Antarctic ice sheets, managed retreats, frequency, climate change, water supplies, adaptation measures, sea level rise, agriculture, homes, weather patterns, sea walls, melting, rising sea levels, global average rate, hydropower generation, global ocean currents, Andes, low-lying islands, glacial retreat, mass loss, coastal erosion, severity, coastal regions, extreme weather events, vulnerable communities
--------------------------------------------------

Step 11 - Node 10 (neighbor of 2):
Content: managed retreats.  
Extreme Weather Events  
Climate change is linked to an increase in the frequenc...
Concepts: Coastal, increased temperatures, frequent droughts, water shortages, changing precipitation patterns, destructive storms, food shortages, ecosystems, water supply, managed retreats, climate change, heatwaves, hurricanes, agriculture, warmer ocean temperatures, heavy rainfall events, economies, urban flooding, resilient infrastructure, droughts, arid regions, communities, storm surge, flooding, intensify, early warning systems, mitigating risks, exacerbate conflicts, heavy rainfall, coastal regions, devastating impacts, extreme weather events
--------------------------------------------------

Complete answer found during traversal.

Final Answer: The main cause of climate change is the increase in greenhouse gases in the atmosphere, primarily driven by human activities such as the burning of fossil fuels (coal, oil, and natural gas) and deforestation. These activities release significant amounts of carbon dioxide (CO2), methane (CH4), and nitrous oxide (N2O), which trap heat from the sun and intensify the greenhouse effect, leading to a warmer climate.

Total Tokens: 15381
Prompt Tokens: 14936
Completion Tokens: 445
Total Cost (USD): $0.0

**Comment by Sandy4321** [comment]
![image](https://github.com/user-attachments/assets/3cb8a73b-94fc-4219-8318-97cb3fae06a0)

**Comment by Sandy4321** [comment]
original
Final Answer: The main cause of climate change is the increase in greenhouse gases in the atmosphere, primarily driven by human activities such as the burning of fossil fuels and deforestation. Greenhouse gases like carbon dioxide (CO2), methane (CH4), and nitrous oxide (N2O) trap heat from the sun, leading to a warmer climate
Mine
Final Answer: The main cause of climate change is the increase in greenhouse gases in the atmosphere, primarily driven by human activities such as the burning of fossil fuels (coal, oil, and natural gas) and deforestation. These activities release significant amounts of carbon dioxide (CO2), methane (CH4), and nitrous oxide (N2O), which trap heat from the sun and intensify the greenhouse effect, leading to a warmer climate.

**Comment by Sandy4321** [comment]
original fig: 
<img width="878" alt="image" src="https://github.com/user-attachments/assets/a653ce25-7448-4dc4-860a-6ef6e2e4a810" />

**Comment by Sandy4321** [comment]
my fig
![image](https://github.com/user-attachments/assets/a1fb94f4-074b-42c9-891a-8526a67eaa85)

**Comment by NirDiamant** [comment]
actually, the next thing to do is to separate dependencies between techniques instead of installing everything, since there are many of them, and not all of them are necessary for all the code tutorials. 
please open issues only on technical nonworking stuff. if you have algorithmic questions, please refer to the discord channel of the project. Nir.

**Comment by Sandy4321** [comment]
I posted several comments on discord channel , but today I found you removed me from your discord channel?
why you do not want to improve your code: make it bugs free and better from theory prospective   ?

**Comment by Sandy4321** [comment]
<img width="1070" alt="image" src="https://github.com/user-attachments/assets/d7c23eda-88cf-4e5b-9f2f-e49046aef662" />
